### PR TITLE
Add support for Vim's tagstack to ALEGoToDefinition

### LINKED ISSUE #1236: Request: add to tag stack when using ALEGoToDefinition OR add a ALEGoBackFromDefinition
Vim shortcuts pair `<C-]> and <C-t>` nicely. With ALEGoToDefinition, there isn't a nice equivalent jump back command (e.g. `<C-t>` ALEGoBackFromDefinition or `<C-t>`) which would be nice to have.

### PR DESCRIPTION
**Pull request by reedriley**
Fixes #1236.

I added tests - but to be fair, this feature won't work on versions of Vim earlier than [8.1.0519](https://github.com/vim/vim/commit/f49cc60aa802862c595ff619dccc11271633a94b) and the test automation script doesn't test on anything that recent.

The feature is designed to degrade gracefully; and I'm open to suggestions on how to best clarify the version requirement in the docs.

### CODE REVIEW

**Comment by reedriley** [comment]
I modified the `Dockerfile` and `./run-tests` script; and there are issues with my proposed unit test.  I'll put up a corrected version shortly.

File: autoload/ale/definition.vim [hunk 1]
```diff
@@ -17,6 +20,18 @@ function! ale#definition#ClearLSPData() abort
     let s:go_to_definition_map = {}
 endfunction
 
+function! ale#definition#UpdateTagStack() abort
+    let l:should_update_tagstack = exists('*gettagstack') && exists('*settagstack') && g:ale_update_tagstack
+
+    if l:should_update_tagstack
+        let l:from = [bufnr('%'), line('.'), col('.'), 0]
```
Comment (w0rp) [inline autoload/ale/definition.vim]: You can use the filename, line, and column we already have to determine this.

File: autoload/ale/definition.vim [hunk 2]
```diff
@@ -17,6 +20,18 @@ function! ale#definition#ClearLSPData() abort
...
+        let l:tagname = expand('<cword>')
+        let l:winid = win_getid()
```
Comment (w0rp) [inline autoload/ale/definition.vim]: Get the window ID from the buffer number, instead of the current window.

File: autoload/ale/definition.vim [hunk 3]
```diff
@@ -17,6 +20,18 @@ function! ale#definition#ClearLSPData() abort
     let s:go_to_definition_map = {}
 endfunction
 
+function! ale#definition#UpdateTagStack() abort
+    let l:should_update_tagstack = exists('*gettagstack') && exists('*settagstack') && g:ale_update_tagstack
+
+    if l:should_update_tagstack
+        let l:from = [bufnr('%'), line('.'), col('.'), 0]
+        let l:tagname = expand('<cword>')
```
Comment (w0rp) [inline autoload/ale/definition.vim]: You might have to wait until the buffer is opened first, and grab the word second.

File: autoload/ale/definition.vim (hunk 1 above)
Comment (reedriley) [reply autoload/ale/definition.vim]: Assuming you're referring to `l:filename`, `l:line` and `l:column` defined in `ale#definition#HandleTSServerResponse` and `ale#definition#HandleLSPResponse` - these are the targets for the new location.  The values I'm pushing on to the stack are based on the old location.

I'll change the variable names to be a bit more clear.  :)

File: autoload/ale/definition.vim (hunk 3 above)
Comment (reedriley) [reply autoload/ale/definition.vim]: Same as above - the values I'm pushing on to the stack are based on the old location.  So I don't think I need to wait for the buffer to open; assuming I've understood this comment correctly?

File: autoload/ale/definition.vim (hunk 2 above)
Comment (reedriley) [reply autoload/ale/definition.vim]: This sounds like it's probably *really* relevant and important - but I'm afraid I'm not an expert on the Vim functions involving windows and buffers.  Any chance you can provide some pointers to the relevant docs?

**Comment by reedriley** [comment]
> We should add tests for this. I'll have to update the Docker image so the 8.1 version we build is new enough to contain this feature.

Agreed.  The test I added is fairly minimal; but passes when I build a custom Docker image and update the run-tests script to use a local image.  I'm happy to include the Dockerfile change as well, if you'd like.

**Comment by GoingFast** [comment]
@w0rp @reedriley doesn't seem to be working for me. :ALEGoToDefinition and when I <C-t> then it says tagstack is empty. So right now I use <C-o>

**Comment by reedriley** [comment]
> @w0rp @reedriley doesn't seem to be working for me. :ALEGoToDefinition and when I C-t then it says tagstack is empty. So right now I use C-o

What version of Vim are you running?

**Comment by wrachwal** [comment]
@w0rp  @reedriley I have the same problem (tagstack is empty). I am using Vim 8.1. Fresh ALE, commit * 27146ade 2019-05-30

**Comment by reedriley** [comment]
@wrachwal: To be clear, I actually meant to ask which version of `Vim`.  The tagstack manipulation functions the diff uses were added in Vim [8.1.0519](https://github.com/vim/vim/commit/f49cc60aa802862c595ff619dccc11271633a94b).

**Comment by wrachwal** [comment]
Thanks, after upgrade using `ppa:jonathonf/vim` it works fine.

**OLD** VIM - Vi IMproved 8.1 (2018 May 18, compiled Nov  6 2018 09:45:53)
Included patches: 1-513

**NEW** VIM - Vi IMproved 8.1 (2018 May 18, compiled May 27 2019 13:46:23)
Included patches: 1-1408

**Comment by dmchoiboi** [comment]
Is there any plan to provide this functionality for NeoVim? I'm on the latest NeoVim version (v0.3.7) and I'm getting the same 'tagstack is empty' mentioned above

**Comment by rislah** [comment]
Yep its not working on neovim.

**Comment by hourliert** [comment]
Same issue here. It is not working on neovim 0.3.8. Can we make it work for neovim by any chance? Thanks in advance!

**Comment by dmchoiboi** [comment]
@hourliert @rislah I just checked NeoVim's source, and it appears that their [patch for vim-8.1.0519](https://github.com/neovim/neovim/pull/9951) hasn't made it out to a stable release yet. According  to that pull request, the update is targeted to be out in the 0.4 release. I've tested out the [nightly release](https://github.com/neovim/neovim/releases/tag/nightly) which contains the fix and verified that this tagstack feature works correctly.

**Pull request merged**